src/main/active_connection.rs

```diff
@@ -267,3 +287,150 @@ enum HeartbeatAction {
+                        let sock = TcpSock::wrap(sock);
+                        unwrap!(el.register(
+                            &sock,
+                            SOCKET_TOKEN,
+                            Ready::writable(),
+                            PollOpt::edge()
+                        ));
+                        socket = Some(sock);
+                    }
+                    SOCKET_TOKEN => {
+                        if let Some(ref mut sock) = socket {
+                            let buffered =
+                                unwrap!(sock.write(Some((Message::Data(vec![1, 2, 3]), 1))));
+                            assert!(buffered);
+                            unwrap!(el.deregister(sock));
+                        }
+                    }
+                    _ => panic!("Unexpected event"),
+                }
+            }
+        }
+    }
+
+    mod active_connection {
+        use super::*;
+
+        mod read {
+            use super::*;
+
+            #[test]
+            fn it_moves_peer_to_the_top_of_the_bootstrap_cache() {
+                let (tx, rx) = mpsc::channel();
+                let _listener_thread = thread::spawn(move || spawn_listener(tx));
+                let listener_port = unwrap!(rx.recv());
+
+                let poll = unwrap!(Poll::new());
+                let sock_token = Token(1);
+                let (peer1_id, _) = rand_peer_id_and_enc_sk();
+                let peer1_sock = unwrap!(TcpSock::connect(&ipv4_addr(127, 0, 0, 1, listener_port)));
+                unwrap!(poll.register(&peer1_sock, sock_token, Ready::writable(), PollOpt::edge()));
+
+                wait_for(sock_token, &poll);
+                let peer1_addr = unwrap!(peer1_sock.peer_addr());
+                let peer1_info = PeerInfo::new(peer1_addr, peer1_id.pub_enc_key);
+
+                let mut cache = bootstrap::Cache::new(Default::default());
+                cache.put(peer1_info);
+                cache.put(peer_info_with_rand_key(ipv4_addr(1, 2, 3, 4, 4000)));
+
+                let mut core = test_core(cache);
+                let (event_tx, _event_rx) = get_event_sender();
+                let (our_id, _) = rand_peer_id_and_enc_sk();
+                let event = Event::ConnectSuccess(peer1_id);
+                ActiveConnection::start(
+                    &mut core,
+                    &poll,
+                    sock_token,
+                    peer1_sock,
+                    our_id,
+                    peer1_id,
+                    CrustUser::Client,
+                    event,
+                    event_tx,
+                );
+
+                let state = unwrap!(core.get_state(sock_token));
+                let mut state = state.borrow_mut();
+                let active_conn = unwrap!(state.as_any().downcast_mut::<ActiveConnection>());
+                // TODO(povilas): debug why wait_for blocks indefinitely on Mac OS
+                //wait_for(sock_token, &poll);
```

Comment:
I don't see why this should fire at all in a deterministic way. You have not registered this socket (`peer1_sock`) with poll to hear about readable readiness so even though the listener writes something, there won't be an notification fired.

Reply:
It's being registered at https://github.com/maidsafe/crust/pull/1139/files#diff-bf742b5e6a9703a78753d7e485151c7dR386

Reply:
@povilasb i don't think i got you. What i said was:
> You have not registered this socket (`peer1_sock`) with poll to hear about **readable** readiness

What you point to has:
```
unwrap!(poll.register(&peer1_sock, sock_token, Ready::writable(), PollOpt::edge()));
```

Can you explain more how is that going to generate a notification if the listener sent something ?

Reply:
I found the problem: it is in the `ActiveConnection::start()` function. This function calls `read()` in its end, before returning, effectively consuming all data in socket's receiving buffer. That means that no `readable` event will ever be raised in this test. It passed previously because of some still unknown reason for `writable` event to be raised again, even though the event was registered as edge-triggered (and was already triggered before).